src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            //Get public properties that have public setter
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                .Where(x => x.GetSetMethod() != null);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                //Base parameters
+                if (algoProperty.PropertyType.BaseType != typeof(AbstractFunction)
+                    && !typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                    if (algoProperty.PropertyType.IsEnum)
+                        parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                    metadata.Parameters.Add(parameter);
+                }
+                //Functions
+                else if (algoProperty.PropertyType.BaseType == typeof(AbstractFunction)
+                         || typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    //Get first public property that is not function base parameters but it inherits it
+                    var functionProperty = algoProperty.PropertyType
+                        .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly |
+                                       BindingFlags.GetProperty |
+                                       BindingFlags.SetProperty) //BindingFlags.DeclaredOnly -> ignore inherited members
+                        .FirstOrDefault(x => x.GetSetMethod() != null &&
+                                             typeof(FunctionParamsBase).IsAssignableFrom(x.PropertyType));
+
+                    //If there is no such property, get property that is function base parameters
+                    if (functionProperty == null)
+                    {
+                        functionProperty = algoProperty.PropertyType
+                            .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.GetProperty |
+                                           BindingFlags.SetProperty)
+                            .FirstOrDefault(x => x.PropertyType == typeof(FunctionParamsBase));
+                    }
+
+                    //If we still cannot find function parameters property, just continue
+                    if (functionProperty == null)
+                        continue;
+
+                    //Check if there is a public constructor which takes FunctionParamsBase
+                    if (!algoProperty.PropertyType.GetConstructors().Any(x =>
+                        x.GetParameters().Any(y => y.ParameterType == functionProperty.PropertyType)))
```

Comment:
This will include constructors with multiple parameters as long as one of them matches `functionProperty.PropertyType`, so you must check the parameter count as well.
It would also probably be better to use `functionProperty.PropertyType.IsAssignableFrom(y.ParameterType)` for the check